### PR TITLE
Support for ghc-9.2

### PR DESCRIPTION
**Pull request by gelisam**
Creating a PR from the branch @junjihashimoto [linked](https://github.com/gelisam/typelevel-rewrite-rules/pull/31#issuecomment-1469051698), to run CI, review the differences, and make a few adjustments.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -70,9 +70,12 @@ jobs:
           # versions of our dependencies allowed by our upper bounds.
           #
           # TODO: switch back to "ghc: latest" once we support ghc-9.0
```
Comment (gelisam) [inline .github/workflows/ci.yml:72]: outdated comment

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -70,9 +70,12 @@ jobs:
           # versions of our dependencies allowed by our upper bounds.
```
Comment (gelisam) [inline .github/workflows/ci.yml:70]: This approach of only testing the oldest and newest supported versions makes sense for most projects, but since this particular project is directly importing the GHC API and that API changes a lot between releases, I think we should test every major GHC version we support.

File: package.yaml [hunk 3]
```diff
@@ -25,10 +25,11 @@ dependencies:
 library:
   source-dirs:      src
   dependencies:
-    - ghc >= 8.10.2 && < 9.0
+    - ghc >= 8.10.2 && <9.4
     - containers >= 0.6.2.1
     - term-rewriting >= 0.3.0.1
     - transformers >= 0.5.6.2
+  default-extensions:  CPP
```
Comment (gelisam) [inline package.yaml:32]: I list the language extensions in every file because some tools (e.g. formatters which transform stdin to stdout) can't look at the cabal file, they can only look at the current file.

File: stack.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
-resolver: nightly-2020-12-10
+resolver: nightly-2021-07-22
```
Comment (gelisam) [inline stack.yaml:1]: ghc-9.0 and ghc-9.2 both have LTSs by now, no need to use a nightly

File: test/error-messages-cases/no-plugin/expected [hunk 5]
```diff
@@ -1 +1,26 @@
...
+        r :: proxy (as ++ (bs ++ cs)) (bound at src/NoPlugin/Test.hs:15:12)
+        ex2e :: proxy as
+                -> proxy bs
+                -> proxy cs
+                -> proxy (as ++ (bs ++ cs))
+                -> proxy ((as ++ bs) ++ cs)
+          (bound at src/NoPlugin/Test.hs:15:1)
+   |
+15 | ex2e _ _ _ r = r
+   |                ^
```
Comment (gelisam) [inline test/error-messages-cases/no-plugin/expected:26]: The new file is suspiciously long, I would restrict it to the relevant part of the error message, like it used to be

File: test/should-compile/GHC/TypeLits/Test.hs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE DataKinds, RankNTypes, TypeFamilies, TypeOperators #-}
+{-# LANGUAGE DataKinds, RankNTypes, TypeFamilies, TypeOperators, AllowAmbiguousTypes #-}
```
Comment (gelisam) [inline test/should-compile/GHC/TypeLits/Test.hs:1]: Why is this needed?

File: test/should-compile/InstanceConstraints/Test.hs [hunk 7]
```diff
@@ -42,7 +42,7 @@ f5 :: forall a b x y
 f5 = foo @(F a x)
 
 f6 :: forall a b x y
-    . ( F a b ~ G x y
+    . ( b ~ G x y
```
Comment (gelisam) [inline test/should-compile/InstanceConstraints/Test.hs:45]: This looks suspicious. If the test has to be changed in order for the test to pass, that means the test is failing, no?

**Comment by gelisam** [comment]
_sigh_, why isn't CI running this time? This is my own PR!!

File: .github/workflows/ci.yml (hunk 2 above)
Comment (gelisam) [reply .github/workflows/ci.yml:70]: And then let's add a "tested-with" line to the cabal file listing those same GHC versions

File: test/should-compile/InstanceConstraints/Test.hs (hunk 7 above)
Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: I see that in ghc-9.2, the unmodified test fails with

```
    • Could not deduce (Foo (F a (G x x))) arising from a use of ‘foo’
      from the context: (F a b ~ G x y, x ~ y, Foo (F a b))
        bound by the type signature for:
                   f6 :: forall a b x y.
                         (F a b ~ G x y, y ~ x, x ~ y, Foo (F a b)) =>
                         ()
        at test/should-compile/InstanceConstraints/Test.hs:(44,1)-(50,8)
    • In the expression: foo @(F a (G x y))
      In an equation for ‘f6’: f6 = foo @(F a (G x y))
```

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: The commit which introduces this test says:

> test that substitutions also work with non-variables
> 
> the code only substitutes variables, but that's good enough because ghc
> replaces nested calls to type families with fsk type variables whose
> definition is in the type substitution.

The body of `f6` is `foo @(F a (G x y))`, which requires a `Foo (F a (G x y))` instance. The constraint `F a b ~ G x y` means that ghc _could_ decide to simplify the requirement to `Foo (F a (F a b))`. This would allow the rewrite rule `forall a b. F a (F a b) ~ F a b` to fire, thus simplifying the requirement to `Foo (F a b)`, which matches the given.

The difficulty is that ghc has no reason to choose to make that "simplification", because `F a b` is not in any way simpler than `G x y`. It is thus the job of typelevel-rewrite-rules to look at the available equality constraints and conclude that `Foo (F a (G x y))` can be simplified to `Foo (F a b)`.

In other words, this is a legitimate test. Changing the constraint from `F a b ~ G x y` to `b ~ G x y` completely changes the test:
1. `b` is simpler than `G x y`, so ghc is more likely to decide to make the simplification on its own, so the test no longer checks that typelevel-rewrite-rules is able to use the available equality constraints to look a few steps ahead.
2. Once ghc performs the simplification, we are left with `Foo (F a b)`, which already matches the given without requiring typelevel-rewrite-rules to rewrite anything.

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: Looking at the contents of `typeSubst`, the problem is clear: ghc-9.2 no longer replaces nested calls to type families with type variables named `fskN`. typelevel-rewrite-rules thus needs to work harder to determine whether rewrite rules should fire.

File: test/should-compile/GHC/TypeLits/Test.hs (hunk 6 above)
Comment (gelisam) [reply test/should-compile/GHC/TypeLits/Test.hs:1]: I figured it out: https://github.com/gelisam/typelevel-rewrite-rules/pull/37#discussion_r1138003124

**Comment by gelisam** [comment]
Merged as part of #37, and #35 is still open in order to track the ghc-9.2 work.

Next, releasing a new version!

File: test/should-compile/GHC/TypeLits/Test.hs (hunk 6 above)
Comment (junjihashimoto) [reply test/should-compile/GHC/TypeLits/Test.hs:1]: Thx!

File: test/should-compile/InstanceConstraints/Test.hs (hunk 7 above)
Comment (junjihashimoto) [reply test/should-compile/InstanceConstraints/Test.hs:45]: Thank you for your review.
The reason why hasktorch didn't generate an error was that this condition was not used by chance.

Comment (junjihashimoto) [reply test/should-compile/InstanceConstraints/Test.hs:45]: I'll see how other projects deal with it.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (gelisam) [reply .github/workflows/ci.yml:72]: Fixed in #37

File: .github/workflows/ci.yml (hunk 2 above)
Comment (gelisam) [reply .github/workflows/ci.yml:70]: Fixed in #37

File: package.yaml (hunk 3 above)
Comment (gelisam) [reply package.yaml:32]: Fixed in #37

File: test/error-messages-cases/no-plugin/expected (hunk 5 above)
Comment (gelisam) [reply test/error-messages-cases/no-plugin/expected:26]: Fixed in #37

File: test/should-compile/GHC/TypeLits/Test.hs (hunk 6 above)
Comment (gelisam) [reply test/should-compile/GHC/TypeLits/Test.hs:1]: Fixed in #37

File: test/should-compile/InstanceConstraints/Test.hs (hunk 7 above)
Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: I can think of two equivalent solutions.

1. do what GHC was doing: replace every sub-expression with a variable. Once we determine that the rewrite can be performed, we must carefully undo those replacements so that we only send GHC a replacement which involves its type variables, not ours.
2. use the unification algorithm (which also allocates one unification variable per sub-expression!) to determine which expression is equal to what. Then, check if any of those type variables unify with the rewrite rule's input. One subtlety is that we can't fail the unification when we encounter mismatched head constructors, because those might be stuck type families, which are allowed to be equal to each other.

To determine which of those two approaches would be the easiest to implement, I need to remind myself how we're already using unification and how we are currently distinguishing between ghc's unification variables and ours.

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: I think we might also need to add a test where the expression to be substituted is inside a larger expression. Previously, with GHC separating all the sub-expressions, it was enough to check if the RHS of a `fskN ~ F a b` constraint matched the rewrite pattern, but it is now necessary to look deeper inside both the LHS and the RHS of `F x (F a b) ~ G x (F a b)`.

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: > I need to remind myself [...] how we are currently distinguishing between ghc's unification variables and ours

The rewrite code is using two different substitution types:

```haskell
type Subst = Map TyVar TypeTerm
type TypeSubst = [(TypeEq, TypeTerm)]
```

Beside the poor naming choice (both hold types, so the "Type" prefix doesn't clarify the difference between the two) and the accidental use of two different data structure, the main difference is the key type: `TyVar` is a type variable, while `TypeEq` can be any type... But the code assumes it's a type variable 🤦

The one which uses `TyVar` keys holds the replacements for the variables in the LHS of a rewrite rule, while the one which uses `TypeEq` keys holds the `fsk1 ~ G x y` constraints from GHC.

The fact that `TypeEq` allows arbitrary types as keys is what allows typelevel-rewrite-rules not to barf on ghc-9.2's `F a b ~ G x y` constraints: the key is now `F a b`. But that doesn't work because if the type expression to be rewritten contains an `F a b`, that part will be expanded to `Fun "F" [Var "a", Var "b"]` before the rewrite begins, and it is only at the leaves that the rewrite code checks whether `"a"` or `"b"` matches the key `F a b`. The substitution is thus not performed, and that's why the test fails on ghc-9.2.

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: Taking a closer look, the data structure choice is not accidental at all! ghc simplifies the constraint `F a b ~ G x y` to _two_ constraints, `fsk1 ~ F a b` and `fsk1 ~ G x y`. Normally an association list with duplicate keys just wastes space because `Data.List.lookup` stops after the first occurrence, but the rewrite code looks at _all_ the possible substitutions. This has exponential cost, but has the benefit of being correct!

Comment (gelisam) [reply test/should-compile/InstanceConstraints/Test.hs:45]: Thinking about this further, there is no reason to expect that `fsk2 ~ H u v` should still hold. If the rewrite rules says that `F a (F a b)` should rewrite to `F a`, that's a claim about a property of `F`, it does not imply that the rewrite should be performed because `F a b ~ F a`. So I now think that the current code is correct, for pre-ghc-9.2 versions at least.

**Comment by sjoerdvisscher** [comment]
Hi, what's the status of this? I'd like to use it with ghc 9.6 and 9.8 :)

**Comment by gelisam** [comment]
> Hi, what's the status of this?

I am not currently working on this, because my main goal was to get to a state where I could merge @junjihashimoto's PR ghc-9.0 PR. Once that [succeeded](https://github.com/gelisam/typelevel-rewrite-rules/pull/31#issuecomment-1471108057), I put this MR on hold on order to focus on other priorities.

> I'd like to use it with ghc 9.6 and 9.8 :)

Alas, this is a type-checker plugin, and is therefore tightly bound to the GHC internals, so some work is needed to _each_ GHC version we want to support. Thus, even if I finished working on this ghc-9.2 branch, it would not help you much for ghc-9.6 nor for ghc-9.8. I can try to get this working with a more recent version of GHC, but probably not with all of ghc-9.2, ghc-9.6, and ghc-9.8. If you could only pick one GHC version, which one would you choose?

**Comment by sjoerdvisscher** [comment]
At the moment both 9.6 and 9.8 work for me. Maybe with a slight preference for ghc 9.8 to be as future proof as possible. The use case is my proarrow library, to make working with monoidal categories, bicategories and double categories as easy as possible.

**Comment by junjihashimoto** [comment]
Hi, when ghc is 9.6, this patch is needed.
 https://github.com/hasktorch/typelevel-rewrite-rules/commit/54d087138afd86aea6de4bc7b9a17e3402978637
